apsuite/commisslib/rfvoltage_calibration.py

```diff
@@ -0,0 +1,126 @@
+
+    def __str__(self):
+        """."""
+        dtmp = '{0:25s} = {1:9d}\n'.format
+        ftmp = '{0:25s} = {1:9.2f}  {2:s}\n'.format
+        stg = ftmp('rfcav_timeout', self.rfcav_timeout, '[s]')
+        stg += ftmp('rfcav_wait', self.rfcav_wait, '[s]')
+        stg += ftmp('rfvoltage_min', self.rfvoltage_min, '[MV]')
+        stg += ftmp('rfvoltage_max', self.rfvoltage_max, '[MV]')
+        stg += dtmp('rfvoltage_nrpoints', self.rfvoltage_nrpoints)
+        stg += ftmp(
+            'conv_physics2hardware', self.conv_physics2hardware, '[mV/MV]')
+        return stg
+
+
+class RFCalibration(_ThreadedMeasBaseClass):
+    """."""
+
+    def __init__(
+            self, isonline=True):
+        """."""
+        super().__init__(target=self._meas_func, isonline=isonline)
+        self.params = RFCalibrationParams()
+        if self.isonline:
+            self.devices['bbbl'] = BunchbyBunch(BunchbyBunch.DEVICES.L)
+            self.devices['rfcav'] = RFCav(RFCav.DEVICES.SI)
+            self.devices['currinfo'] = CurrInfoSI()
+
+    def calc_rfvoltage_span(self):
+        """."""
+        prms = self.params
+        rfvoltage_span = _np.linspace(
+            prms.rfvoltage_min, prms.rfvoltage_max, prms.rfvoltage_nrpoints)
+        return rfvoltage_span
+
+    def configure_bbb_drive_pattern(self, cbmode):
+        """."""
+        harm_nr = 864
+        bunches = _np.arange(harm_nr)
+        bbbl = self.devices['bbbl']
+        bbbl.drive0.mask = _np.cos(2*_np.pi*bunches*cbmode/harm_nr) > 0
+        # drive at synchrotron frequency
+        bbbl.drive0.frequency = bbbl.sram.modal_marker_freq
+
+    def _meas_func(self):
+        data = dict()
+        props = [
+            'timestamp', 'sync_freq', 'sync_tune', 'rf_frequency',
+            'stored_current', 'rfvoltage_rb', 'rfamplitude_rb']
+        for key in props:
+            data[key] = []
+
+        rfcav, bbbl = self.devices['rfcav'], self.devices['bbbl']
+
+        amp0 = rfcav.dev_llrf.voltage
+        prms = self.params
+
+        rfamp_span = self.calc_rfvoltage_span()
+        rfamp_span *= prms.conv_physics2hardware
+
+        data['rfvoltage_sp'] = rfamp_span/prms.conv_physics2hardware
+        data['rfamplitude_sp'] = rfamp_span
+
+        # set first value of voltage
+        rfcav.set_voltage(rfamp_span[0], timeout=prms.rfcav_timeout)
+        for amp in rfamp_span:
+            if self._stopevt.is_set():
+                print('Stopping...')
+                break
+
+            rfcav.set_voltage(amp, timeout=prms.rfcav_timeout)
+            _time.sleep(prms.rfcav_wait)
+            sync_freq = bbbl.sram.modal_marker_freq
+            sync_tune = bbbl.sram.modal_marker_tune
+            rffreq = rfcav.dev_rfgen.frequency
+            scurr = self.devices['currinfo'].current
+            gap_volt = rfcav.dev_cavmon.gap_voltage
+            amp_volt = rfcav.dev_llrf.voltage
+
+            data[props[0]].append(_time.time())
```

Comment:
this indirect reference to the properties may lead to bugs in the future, why not using the string directly?